Document path resolution

raddevon opened the pull request:
Addresses #4441

zchenb commented on docs/edgeql/path_resolution.rst:
```diff
@@ -0,0 +1,151 @@
+use the same *symbol* to refer to a set. This means operations with
+``User.first_name`` and ``User.last_name`` do get the common path factored
+while ``U.first_name`` and ``User.last_name`` do not and are resolved with
+cartesian multiplication.
```
I am not sure but it seems that there are better explanation, that the clauses of with and the body of with live in different scopes (same reason as for the next section). Maybe Sully can confirm?

zchenb replied:
The technical reason is that `U` in `U.first_name` is really a binder that has not been instantiated. Putting it on a par with `User.last_name` seems strange.

raddevon replied:
Any ideas for a better way to explain this @vpetrovykh or @msullivan? This was my takeaway from my conversation with Victor. I may have misunderstood.

msullivan replied:
Both of these reasons are correct and are part of what is going on:
 * `U.first_name` and `User.last_name` don't have anything factored out because they don't have a common prefix
 * The `User` in the `with` binding and the `User` in the `select` body don't get factored because they are in separate subqueries

raddevon replied:
Makes sense. I will update it to reflect that. That will give me a natural segue into scopes too. Thanks @zchenb and @msullivan!

zchenb reviewed (commented):
Looks fine to me as an introductory explanation for path resolution except the commented line. The text did not mention the interaction with subqueries and semi-subqueries which are more advanced topics.

msullivan commented on docs/edgeql/path_resolution.rst:
```diff
@@ -0,0 +1,151 @@
+
+We assume this is what you want, but if your goal is to get the cartesian
+product, you can accomplish it one of two ways. You could use
+:eql:op:`detached`.
```
There's a third way, too, which is to put both parts in a subquery like in the "Scopes" section

msullivan commented on docs/edgeql/path_resolution.rst:
```diff
@@ -0,0 +1,151 @@
+      ('John', 'Seward'),
+    }
+
+Common symbols in nested scopes *are* factored when they use the same symbol.
```
I think another good example here is `select (User.name, count(User.friends))`, which returns each user's name and the number of friends they have.

raddevon commented:
I believe I have addressed all of the feedback in this latest push. @msullivan and/or @zchenb, would you care to have another look when you have time?

zchenb commented on docs/edgeql/path_resolution.rst:
```diff
@@ -0,0 +1,177 @@
+Scopes
+------
+
+Scopes change the was path resolution works. Two sibling select queries — that
```
`Scopes change the was path resolution works.` Not sure how to read this sentence. Could you clarify?

raddevon replied:
That's a typo. Thanks for catching it! Just pushed a fix. Should be `way` instead of `was`.

vpetrovykh commented on docs/edgeql/path_resolution.rst:
```diff
@@ -0,0 +1,177 @@
+
+.. code-block:: edgeql-repl
+
+    edgedb> with U := (select User)
```
```suggestion
    edgedb> with U := User
```
You don't need the `select` for aliasing simple expressions. You usually need `select` if you want to use some of the other clauses (`filter` or `order`, etc.).

vpetrovykh commented on docs/edgeql/path_resolution.rst:
```diff
@@ -0,0 +1,177 @@
+``Person`` objects that are in *all* ``friends`` links rather than the number
+of ``Person`` objects that are in the named ``Person`` object's ``friends``
+link.
+
```
It's possibly worthwhile contrasting this with an example where both parts of that tuple are wrapped into some aggregate function:
```
    edgedb> select (array_agg(distinct Person.name), count(Person.friends));
    {(['Fran', 'Bam', 'Emma', 'Geoff'], 3)}
```
Here `Person` is not factored out because the scopes of aggregate function arguments are _sibling_ scopes, much like the two `select` sub-queries from earlier. So the interpretation is "get an array of all unique names" and "get the total number of people who are somebody's friend".

raddevon replied:
Good suggestion. I added this (nearly verbatim).

vpetrovykh commented on docs/edgeql/path_resolution.rst:
```diff
@@ -0,0 +1,177 @@
+common symbols are factored and assumed to refer to the same object as the
+outer query. This is because clauses like :ref:`filter
+<ref_eql_select_filter>` and :ref:`order by <ref_eql_select_order>` need to
+be applied to each row.
```
```suggestion
be applied to each value in the result.
```
We generally avoid using "row" or "column" in our explanations as we don't actually have them.

vpetrovykh commented on docs/edgeql/path_resolution.rst:
```diff
@@ -0,0 +1,177 @@
+be applied to each row.
+
+The :ref:`limit <ref_eql_select_pagination>` clause is not nested in the scope
+because it needs to be applied globally to your query.
```
```suggestion
because it needs to be applied globally to the entire result set of your query.
```

vpetrovykh reviewed (approved):
LGTM
